src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -25,121 +25,113 @@ private sealed class ChunkedEncodingReadStream : HttpContentReadStream
-                if (!Utf8Parser.TryParse(line.AsReadOnlySpan(), out ulong chunkSize, out int bytesConsumed, 'X'))
+                // Try to consume from data we already have in the buffer.
+                int bytesRead = ReadChunksFromConnectionBuffer(destination.Span);
+                if (bytesRead > 0)
                 {
-                    throw new IOException(SR.net_http_invalid_response);
+                    return new ValueTask<int>(bytesRead);
                 }
-                else if (bytesConsumed != line.Count)
+
+                // Nothing available to consume.  Fall back to I/O.
+                return ReadAsyncCore(destination, cancellationToken);
+            }
+
+            private async ValueTask<int> ReadAsyncCore(Memory<byte> destination, CancellationToken cancellationToken)
+            {
+                // Should only be called if ReadChunksFromConnectionBuffer returned 0.
+
+                Debug.Assert(_connection != null);
+                Debug.Assert(destination.Length > 0);
+
+                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
+                try
                 {
-                    // There was data after the chunk size, presumably a "chunk extension".
-                    // Allow tabs and spaces and then stop validating once we get to an extension.
-                    int offset = line.Offset + bytesConsumed, end = line.Count - bytesConsumed + line.Offset;
-                    for (int i = offset; i < end; i++)
+                    while (true)
                     {
-                        char c = (char)line.Array[i];
-                        if (c == ';')
+                        if (_state == ParsingState.Ending)
                         {
-                            break;
+                            await ConsumeTrailingHeaders().ConfigureAwait(false);
+                            Finish();
+                            return 0;
                         }
-                        else if (c != ' ' && c != '\t') // not called out in the RFC, but WinHTTP allows it
+
+                        // We're only here if we need more data to make forward progress.
+                        await _connection.FillAsync();
+
+                        // Now that we have more, see if we can get any response data, and if
+                        // we can we're done.
+                        int bytesCopied = ReadChunksFromConnectionBuffer(destination.Span);
+                        if (bytesCopied > 0)
                         {
-                            throw new IOException(SR.net_http_invalid_response);
+                            return bytesCopied;
                         }
                     }
                 }
-
-                _chunkBytesRemaining = chunkSize;
-                if (chunkSize > 0)
+                catch (Exception exc) when (ShouldWrapInOperationCanceledException(exc, cancellationToken))
```

Comment:
Not for this PR, but:

I wonder if it would be better to have this exception translation logic in places like FillAsync where we are actually doing the operations on the underlying stream.  There are only a handful of these I think, and it help ensure that we are doing the exception translation consistently.  Thoughts?

Not sure about the perf impact; my guess is it's irrelevant.

Reply:
We want the try/finally at this level so that we don't register/unregister with the cancellation token on every individual filling of the buffer.  As long as we have the try/finally, seems like we may as well have the try/catch/finally in order to keep all of the cancellation-related logic in the same place, avoid another try/catch in addition to the try/finally, etc.